File: Java/MergeSort.java
class Sort{

    public void merge(int[] arr,int left,int mid,int right){
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] arr1 = new int[n1];
        int[] arr2 = new int[n2];

        for (int i = 0; i < n1; i++){
            arr1[i] = arr[left + i];
        }
            
        for (int j = 0; j < n2; j++){
            arr2[j] = arr[mid + 1 + j];
        }
            
        int i = 0;
        int j = 0;

        int k = left;

        while(i < n1 && j < n2){
            if(arr1[i] < arr2[j]){
                arr[k] = arr1[i];
                i++;
            }

            else{
                arr[k] = arr2[j];
                j++;
            }

            k++;
        }

        while(i < n1){
            arr[k] = arr1[i];
            i++;
            k++;
        }

        while(j < n2){
            arr[k] = arr2[j];
            j++;
            k++;
        }

    }


    public void sort(int[] arr,int left,int right){
        if(left < right){
            int mid = (left + right)/2;

            sort(arr,left,mid);
            sort(arr,mid+1,right);

            merge(arr,left,mid,right);
        }
    }
}

public class MergeSort{
    public static void main(String[] args) {
        int[] arr = {2,3,1,4,6,5};

        System.out.print("Initial Array Before Sorting : ");

        for(int i = 0;i < arr.length;i++){
            System.out.print(arr[i] + " ");
        }

        System.out.println(" ");

        Sort s = new Sort();

        s.sort(arr, 0, arr.length-1);

        System.out.print("Sorted Array Using MergeSort : ");

        for(int i = 0;i < arr.length;i++){
            System.out.print(arr[i] + " ");
        }

        System.out.println(" ");

    }
}

File: c++/sorts/mergesort.cpp
// MERGE SORT 
// https://www.geeksforgeeks.org/merge-sort/
//TIME COMPLEXCITY = O(n log n)

#include<iostream>
using namespace std;

void merge(int *arr, int s, int e){

 int mid = (s+e)/2;

//length of both
int len1 = mid-s+1;
int len2 = e - mid;

int *first = new int[len1];
int *second = new int[len2];

int mainArrayIndex= s;
for (int i = 0; i < len1; i++)
{
    first[i]=arr[mainArrayIndex++];
}

mainArrayIndex= mid+1;
for (int i = 0; i < len2; i++)
{
    second[i]=arr[mainArrayIndex++];
}

//merge 2 sorted array

int index1 = 0;
int index2 = 0;
mainArrayIndex = s;
while (index1 < len1 && index2 < len2)
{
    if (first[index1]<second[index2])
    {
        arr[mainArrayIndex++] = first[index1++];
    }
    else
    arr[mainArrayIndex++] = second[index2++];
}

while (index1<len1)
{
     arr[mainArrayIndex++] = first[index1++];
}
while (index2<len2)
{
     arr[mainArrayIndex++] = second[index2++];
}


}

void mergeSort(int *arr, int s , int e){
    //base case 
    if (s>=e)
    {
        return;
    }
    int mid = (s+e)/2;

    //left part
    mergeSort(arr,s,mid);

    //right part
    mergeSort(arr, mid+1 , e);

    //merge both the parts 
    merge(arr,s,e);

}

int main()
{
    int arr[5] = {2,5,9,3,1};
    int n= 5;

    mergeSort(arr , 0 , n-1);
    for (int i = 0; i < n; i++)
    {
        cout<< arr[i] <<" "<<endl;
    }
    
// cout<< arr <<" "<<endl;
return 0;
}




